test/jdk/tools/jar/ValidatorTest.java

```diff
@@ -0,0 +1,334 @@
+        }
+    }
+
+    @Test
+    public void testOnlyLocModified() throws IOException {
+        Path f = Path.of("LocHacked.jar");
+        writeManifestAsFirstSecondAndFourthEntry(f, false, true);
+        try {
+            jar("--validate --file " + f.toString());
+            fail("Expecting non-zero exit code");
+        } catch (IOException e) {
+            var err = e.getMessage();
+            System.out.println(err);
+            assertTrue(err.contains("Warning: There were 2 local file headers found for META-INF/MANIFEST.MF"));
+            assertTrue(err.contains("Warning: An equivalent for the central directory entry META-INF/AANIFEST.MF was not found in the local file headers"));
+            // Order is base on the central directory, expecting AANIFEST.MF but see next entry
+            assertTrue(err.contains("Warning: Central directory and local file header entries are not in the same order"));
+        }
+    }
+
+    @Test
+    public void testOnlyCenModified() throws IOException {
+        Path f = Path.of("CenHacked.jar");
+        writeManifestAsFirstSecondAndFourthEntry(f, true, false);
+        try {
+            jar("--validate --file " + f.toString());
+            fail("Expecting non-zero exit code");
+        } catch (IOException e) {
+            var err = e.getMessage();
+            System.out.println(err);
+            assertTrue(err.contains("Warning: There were 3 central directory entries found for META-INF/MANIFEST.MF"));
+            assertTrue(err.contains("Warning: An equivalent entry for the local file header META-INF/AANIFEST.MF was not found in the central directory"));
+            assertTrue(err.contains("Warning: An equivalent entry for the local file header META-INF/BANIFEST.MF was not found in the central directory"));
+            assertFalse(err.contains("Warning: Central directory and local file header entries are not in the same order"));
+        }
+    }
+
+    @Test
+    public void testMismatchOrder() throws IOException {
+        Path f = Path.of("SwappedEntry.jar");
+        createMismatchOrderJar(f);
+        try {
+            jar("--validate --file " + f.toString());
+            fail("Expecting non-zero exit code");
+        } catch (IOException e) {
+            var err = e.getMessage();
+            assertTrue(err.contains("Warning: Central directory and local file header entries are not in the same order"));
+        }
+    }
+
+    @Test
+    public void testInvalidEntryName() throws IOException {
+        Path f = Path.of("InvalidEntry.jar");
+        var invalidEntryNames = createInvalidEntryJar(f);
+        try {
+            jar("--validate --file " + f.toString());
+            fail("Expecting non-zero exit code");
+        } catch (IOException e) {
+            var err = e.getMessage();
+            System.out.println(err);
+            for (var entryName : invalidEntryNames) {
+                assertTrue(err.contains("Warning: entry name " + entryName + " is not valid"));
```

Comment:
To help debug any failures, it might be better to include the entryName that failed the assert, something like:

```java
assertTrue(err.contains("Warning: entry name " + entryName + " is not valid"), "missing warning for " + entryName);
```